src/compiler/crystal/interpreter/multidispatch.cr

```diff
@@ -168,8 +168,21 @@ module Crystal::Repl::Multidispatch
       call_args = [] of ASTNode
 
       i = 0
-      node.args.each do
-        call_args << Var.new("arg#{i}")
+      node.args.each_with_index do |arg, arg_index|
+        var = Var.new("arg#{i}")
+
+        # If the argument was autocasted it will always match in a multidispatch
+        if autocast_types.try &.[arg_index]?
+          call_args << var
+          next
+        end
+
+        target_def_arg = target_def.args[i]
+
+        # Make sure to always cast the argument to the target def arg's type
+        cast = Cast.new(var, TypeNode.new(target_def_arg.type))
```

Comment:
Just wondering, could we skip this cast if arg type and target type are identical?

Reply:
That's true! 

I'm now thinking that we only need the cast in the last `if`, where there's no restriction on the argument. Because otherwise the arg's type is always checked. Let me try it...

Reply:
Done!

Reply:
To expand a bit more on this: a multidispatch will generate code like this:

```crystal
def multidispatch(arg)
  if arg.is_a?(String)
    foo(arg)
  elsif arg.is_a?(Int32)
    foo(arg)
  else
    foo(arg)
  end
end
```

In this case if `arg`'s type is `String | Int32 | Nil` the last branch will have `arg`'s type to be `Nil`. But, the method corresponding to `foo` might either take `Nil` or maybe `String | Nil`, or even `String | Int32 | Nil` depending on how the overloads are defined.

To make the cast conditionally on the last branch we'd have to know what the type of `arg` is there: removing any types that were already checked in `.is_a?`. Given that this is an optimization, I'd like to leave it for later or maybe not even implement it later (at most we just have one Cast node that does nothing)